### PR TITLE
Tree: implement delta application for object-forest

### PR DESCRIPTION
**Pull request by CraigMacomber**

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 1]
```diff
@@ -321,6 +377,16 @@ class Cursor implements ITreeSubscriptionCursor {
         return this.nodeStack[this.nodeStack.length - 1];
     }
 
+    getParentFieldKey(): FieldKey | DetachedField {
+        assert(this.state === ITreeSubscriptionCursorState.Current, 0x33d /* Cursor must be current to be used */);
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts]: The build is complaining about this error code being copied from line 375.

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts [hunk 2]
```diff
@@ -38,13 +37,114 @@ export class ObjectForest extends SimpleDependee implements IEditableForest {
...
+            },
+            onMoveOut: (index: number, count: number, id?: Delta.MoveId): void => {
+                assert(currentField !== undefined, "must be in field to onMoveOut");
+                let srcField: ObjectField;
+                if (currentNode.state === ITreeSubscriptionCursorState.Cleared) {
+                    srcField = this.getRoot(currentField as unknown as DetachedField);
+                } else {
+                    srcField = getGenericTreeField(currentNode, currentField, false);
+                }
+                const field = this.detachRangeOfChildren(srcField, index, index + count);
```
Comment (alex-pardes) [inline packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:95]: How do we ensure that the DetachedField created here doesn't collide with a move destination referenced in the delta?

**Review by alex-pardes — APPROVED** [review]

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts (hunk 1 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts]: fixed

File: packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts (hunk 2 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:95]: they are separate scopes. In here we never use move ids as detached fields or field keys, we instead allocate a new unique detached field for each move id and store them in the `moves` map.

Comment (alex-pardes) [reply packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts:95]: On line 60 we interpret a FieldKey specified by the delta as a DetachedField, and then edit that field. Even if we assume that there were no detached fields at the start of applying the delta, isn't there a risk that the specified detached field was previously created to be used to temporarily store moved out content?

**Pull request merged**